semgrep/semgrep/output.py

```diff
@@ -173,3 +187,178 @@ def build_output(
+    ):
+        self.settings = output_settings
+        self.stderr = stderr
+        self.stdout = stdout
+
+        self.rule_matches: List[RuleMatch] = []
+        self.debug_steps_by_rule: Dict[Rule, List[Dict[str, Any]]] = {}
+        self.rules: FrozenSet[Rule] = frozenset()
+        self.semgrep_core_errors: List[Dict[str, Any]] = []
+        self.semgrep_rule_errors: List[SemgrepError] = []
+        self.has_output = False
+
+        self.exit_code = 0
+        self.final_error: Optional[Exception] = None
+
+    def handle_semgrep_core_errors(self, semgrep_errors: List[Dict[str, Any]]) -> None:
+        """
+        Report errors coming directly from semgrep-core (raw JSON objects)
+        """
+        self.semgrep_core_errors += semgrep_errors
+        if self.settings.output_format == OutputFormat.TEXT:
+            for error in semgrep_errors:
+                print_error(pretty_error(error))
+
+    def handle_semgrep_rule_errors(self, error: SemgrepError) -> None:
+        """
+        Report parse errors from semgrep rules. Either:
+        - when the pattern fails to parse in semgrep-core
+        - when the YAML or YAML structure is invalid
+        """
+        self.semgrep_rule_errors.append(error)
+        self._output_exception(error)
+
+    def handle_semgrep_core_output(
+        self,
+        rule_matches_by_rule: Dict[Rule, List[RuleMatch]],
+        debug_steps_by_rule: Dict[Rule, List[Dict[str, Any]]],
+    ) -> None:
+        self.has_output = True
+        self.rules = self.rules.union(rule_matches_by_rule.keys())
+        self.rule_matches += [
+            match
+            for matches_of_one_rule in rule_matches_by_rule.values()
+            for match in matches_of_one_rule
+        ]
+
+        self.debug_steps_by_rule.update(debug_steps_by_rule)
+
+    def handle_unhandled_exception(self, ex: Exception) -> None:
+        """
+        This is called by the context manager upon an unhandled exception. If you want to record a final
+        error & set the exit code, but keep executing to perform cleanup tasks, call this method.
+        """
+        if isinstance(ex, SemgrepError):
+            self.exit_code = ex.code
+        else:
+            self.exit_code = FATAL_EXIT_CODE
+        self.final_error = ex
+
+    def _output_exception(self, ex: Exception) -> None:
+        if isinstance(ex, SemgrepError):
+            print_error(str(ex))
+        else:
+            # If it isn't a known SemgrepError, bail hard.
+            print_error(PLEASE_FILE_ISSUE_TEXT)
+            raise ex
+
+    def close(self) -> int:
+        """
+        Close the output handler.
+
+        This will write any output that hasn't been written so far. It returns
+        the exit code of the program.
+        """
+        # TODO: incorporate final_error into JSON output (https://github.com/returntocorp/semgrep/issues/746)
+        if self.final_error:
+            self._output_exception(self.final_error)
```

Comment:
Instead of dealing with something like `final_error` we could just continue to catch exceptions in `__main__` and output their contents there. If you still want to use something like `_output_exception` you could make a `@staticmethod` (it doesn't use `self` anyway) and use it in `__main__`.

Again, it feels kinda weird to have our output handler also manage all program exceptions as well. What if we want to catch and handle them elsewhere?

Reply:
if you want to capture an exception internally, handle it an move on, there is no reason you can't do that. The output manager handles exceptions to provide a unified way of printing them and incorporating them into JSON output.

In `__main__` we don't actually know anything about the output mode, so we don't have the context we need to know how to handle the exception.